### PR TITLE
fix(transform_processor): Complete fully filtered transform inputs without emitting empty PData

### LINKED ISSUE #3891: Empty transform output can enter a durable retry loop
### Pre-filing checklist

- [x] I searched existing issues and didn't find a duplicate

### Component(s)

Rust OTAP dataflow (rust/otap-dataflow/)

### Bug Description

When an OPL transform filters every record from an input PData message, the transform processor forwards a zero-item PData message through its default output instead of completing the input without downstream output.

This creates observable downstream work for a message with no remaining telemetry. In a pipeline with a durable buffer, the empty message is persisted as a real Quiver bundle. If a downstream component transiently NACKs empty input, that bundle is retried unchanged until retention expiry.

```text
one input item
  -> transform: logs | where false
  -> zero-item PData forwarded
  -> durable buffer persists empty bundle
  -> transient downstream NACK
  -> repeated retries until max_age
```

The dedicated filter processor already handles this case by completing and ACKing zero-item output without forwarding it. The transform processor should provide equivalent completion behavior when its default and routed outputs are all empty.

### Steps to Reproduce

1. Save the configuration below as `transform-empty-output.yaml`.
2. From `rust/otap-dataflow`, remove any state from a previous run:

   ```bash
   rm -rf /tmp/otap-transform-empty-output-repro
   ```

3. Start a single-core dataflow:

   ```bash
   cargo run --release --bin df_engine -- --config /absolute/path/to/transform-empty-output.yaml --num-cores 1
   ```

4. Observe that the durable buffer forwards a bundle even though `logs | where false` removed the only generated log.
5. Observe the same zero-item bundle being transiently NACKed and retried.
6. After segment finalization and the five-second `max_age`, observe one expired bundle with zero expired items.

The error exporter is intentional: unlike the noop exporter, it exposes that an empty message crossed the transform boundary by keeping the resulting durable bundle unresolved. Replacing it with `exporter:noop` ACKs the empty message immediately and hides the retry/expiry symptom.

### Expected Behavior

- The transform records one filtered log item.
- Because no default or routed output contains data, the processor completes and ACKs the inbound message without sending downstream PData.
- The durable buffer receives no message, persists no bundle, performs no retries, and reports no expiry.

### Actual Behavior

- The transform reports successful execution after filtering the only item.
- It sends a zero-item PData message through the default output.
- The durable buffer persists and forwards that message as a real bundle.
- The error exporter transiently NACKs it.
- The durable buffer repeatedly retries the unchanged empty bundle until it expires.

Expected diagnostic values from the reproduction are:

```text
repro_transform_operations{signal="logs",outcome="success"} = 1
repro_durable_retries_scheduled > 0
repro_durable_bundle_outcomes{outcome="deferred"} > 0

repro_durable_lost_segments{reason="expired"} = 1
repro_durable_lost_bundles{reason="expired"} = 1
repro_durable_lost_items{signal="logs",reason="expired"} = 0
```

The zero item-loss value is expected in this isolated reproduction: the defect is that an empty PData envelope becomes downstream durable work at all.

### OTel-Arrow Version

latest

### Environment

N/A

### Configuration

```yaml
version: otel_dataflow/v1

engine:
  http_admin:
    bind_address: 127.0.0.1:18080
  telemetry:
    reporting_interval: 1s
    logs:
      level: debug
  observability:
    pipeline:
      policies:
        telemetry:
          pipeline_metrics: false
          tokio_metrics: false
          runtime_metrics: none
      nodes:
        telemetry:
          type: receiver:internal_telemetry
          config:
            signals: [metrics]
            metrics:
              interval: 1s
              views:
                - selector:
                    scope_name: processor.transform
                    instrument_name: operations
                  stream:
                    name: repro_transform_operations
                - selector:
                    scope_name: processor.durable_buffer
                    instrument_name: retries.scheduled
                  stream:
                    name: repro_durable_retries_scheduled
                - selector:
                    scope_name: processor.durable_buffer.bundles
                    instrument_name: resolved
                  stream:
                    name: repro_durable_bundle_outcomes
                - selector:
                    scope_name: processor.durable_buffer.loss
                    instrument_name: segments
                  stream:
                    name: repro_durable_lost_segments
                - selector:
                    scope_name: processor.durable_buffer.loss
                    instrument_name: bundles
                  stream:
                    name: repro_durable_lost_bundles
                - selector:
                    scope_name: processor.durable_buffer.loss
                    instrument_name: items
                  stream:
                    name: repro_durable_lost_items
        metric_filter:
          type: processor:filter
          config:
            metrics:
              include:
                match_type: strict
                metric_names:
                  - repro_transform_operations
                  - repro_durable_retries_scheduled
                  - repro_durable_bundle_outcomes
                  - repro_durable_lost_segments
                  - repro_durable_lost_bundles
                  - repro_durable_lost_items
        metrics_debug:
          type: processor:debug
          config:
            verbosity: detailed
            mode: batch
            signals: [metrics]
        telemetry_sink:
          type: exporter:noop
      connections:
        - from: telemetry
          to: metric_filter
        - from: metric_filter
          to: metrics_debug
        - from: metrics_debug
          to: telemetry_sink

groups:
  reproduction:
    pipelines:
      main:
        policies:
          channel_capacity:
            control:
              node: 100
              pipeline: 100
            pdata: 16
        nodes:
          generator:
            type: receiver:traffic_generator
            config:
              data_source: synthetic
              traffic_config:
                production_mode: open
                max_signal_count: 1
                max_batch_size: 1
                signals_per_second: 1
                metric_weight: 0
                trace_weight: 0
                log_weight: 100
          drop_all:
            type: processor:transform
            config:
              opl_query: logs | where false
              skip_sanitize_result: true
          durable_buffer:
            type: processor:durable_buffer
            config:
              path: /tmp/otap-transform-empty-output-repro
              poll_interval: 100ms
              max_age: 5s
              retention_size_cap: "192 MiB"
              size_cap_policy: backpressure
              max_segment_open_duration: 1s
              initial_retry_interval: 250ms
              max_retry_interval: 1s
          transient_nack:
            type: exporter:error
            config:
              message: deterministic downstream rejection
        connections:
          - from: generator
            to: drop_all
          - from: drop_all
            to: durable_buffer
          - from: durable_buffer
            to: transient_nack
```

### Log Output

```shell

```

### Additional Context

The production trigger was a generated receiver-format transform that intentionally removed an unsupported input variant. Whether the rejected record became an isolated empty PData depended on receiver batching, which made the downstream symptom intermittent even though the filter decision was deterministic.

### PR DESCRIPTION
**Pull request by drewrelmas**
# Change summary

- complete fully filtered transform inputs without emitting empty PData
- suppress empty default and routed outputs after sanitization
- preserve upstream ACK/NACK completion when only routed outputs contain data

The transform processor can emit a default output plus named outputs created by `route_to`. This change forwards only outputs containing semantic telemetry. If all outputs are empty, it completes and ACKs the input locally. If only named outputs contain data, it omits the empty default output and lets those named outputs control upstream completion through the existing split-context tracking.

| Default output | Named outputs | Behavior |
|---|---|---|
| Non-empty | None | Forward the default output directly |
| Empty | None | Complete locally and ACK |
| Non-empty | Non-empty | Forward and track every non-empty output |
| Empty | Non-empty | Omit the default output; named outputs control completion |
| Any | Empty named outputs | Suppress the empty named outputs |

## Related issue

* Fixes #3891 
  * Note, this is a targeted fix for this instance, but worth larger discussion to see if we should introduce an invariant to handle future cases.

## Validation

Unit tests

## User-facing changes

The `transform_processor` no longer forwards empty batches to downstream components.

### CODE REVIEW

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 1]
```diff
@@ -1039,6 +1058,40 @@ mod test {
             });
     }
 
+    #[test]
+    fn test_fully_filtered_transform_completes_without_output() {
```
[the remark this answers is not in the document]
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs:1074]: Addressed

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 2]
```diff
@@ -291,25 +292,63 @@ impl TransformProcessor {
...
+                .connected_ports()
+                .iter()
+                .find(|p| p.as_ref() == route_name.as_str())
+                .ok_or_else(|| {
+                    TransformOperationError::new(
+                        TransformErrorType::RouteNotConfigured,
+                        EngineError::ProcessorError {
+                            processor: effect_handler.processor_id(),
+                            kind: ProcessorErrorKind::Transport,
+                            error: "Routing error: ".into(),
```
[the remark this answers is not in the document]
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Addressed

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 3]
```diff
@@ -330,91 +369,71 @@ impl TransformProcessor {
...
+        if default_has_data {
+            let mut pdata = OtapPdata::new(outbound_context.clone(), default_otap_batch.into());
+            let outbound_key = self
+                .contexts
+                .insert_outbound(inbound_ctx_key)
+                .ok_or_else(|| {
+                    // if we can't emit the default batch, we won't be able to route any of the
+                    // routed batches, so clear them to ensure they're not stuck in the router's
+                    // buffer
+                    router_impl.routed.clear();
```
[the remark this answers is not in the document]
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Old iteration

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 4]
```diff
@@ -289,27 +290,72 @@ impl TransformProcessor {
...
-        // non-empty batch to be output, we don't need to change any contexts or subscriptions
+        // Empty named outputs must not consume downstream capacity or participate
+        // in the inbound request's ACK/NACK completion count.
+        router_impl
+            .routed
+            .retain(|(_, routed_batch)| !routed_batch.is_empty());
+        let default_has_data = !default_otap_batch.is_empty();
 
+        // With no named outputs, preserve the direct-context fast path. A fully
+        // empty transform instead completes the original request locally.
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: nit: "fully empty transform" reads like there was a transformation with no steps. I feel that this comment should indicate that it is the batch that is empty.

```suggestion
        // With no named outputs, preserve the direct-context fast path. A fully
        // empty transform result instead completes the original request locally.
```

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 5]
```diff
@@ -289,27 +290,72 @@ impl TransformProcessor {
         // Record flow metrics from the engine's per-batch execution counters
         effect_handler.record_flow_dropped_items(signal, counters.filtered as u64);
 
+        // Normalize every candidate output before deciding whether it contains
+        // semantic telemetry. This keeps default and named outputs consistent.
         if self.sanitize_results {
             sanitize_otap_batch(&mut default_otap_batch);
+            for (_, routed_batch) in &mut router_impl.routed {
+                sanitize_otap_batch(routed_batch);
+            }
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: I see that alongside the empty batch fixed, we also pre-sanitize and pre-resolve the port_names. Is there a reason for that? Is it because when we encountered an error doing either of these operations, in the old code, we'd have already allocated contexts that would be stuck in the slot map?

I'm not opposed to this, but it does mean we have an extra allocation for the `routed_outputs` on line 335, which is probably not a huge deal in the grand scheme of things.

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 6]
```diff
@@ -289,27 +290,72 @@ impl TransformProcessor {
...
         }
 
-        // TODO - there's probably some optimization we can make below where if there's only one
-        // non-empty batch to be output, we don't need to change any contexts or subscriptions
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs:297]: nit: maybe keep this TODO. The intention here was that, if there was no default output but there was only a single routed output, then we could just reuse the inbound context for the one routed output.

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs [hunk 7]
```diff
@@ -330,91 +376,70 @@ impl TransformProcessor {
...
+                .send_message_with_source_node(pdata)
+                .await
+                .map_err(|error| {
+                    TransformOperationError::new(TransformErrorType::OutputSend, error.into())
+                })?;
+        }
 
+        // handle any batches that need to be forwarded to a specific output port thanks to invocation
+        // of a "route_to" operator call
+        for (port_name, otap_batch) in routed_outputs {
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Maybe I've missed it, but are we checking anywhere that the routed batches are not empty as well?  e.g. if we had a transform like this, we'd end up with a routed empty batch.
```
logs | where severity_text = "something invalid" | route_to "out_port"
```

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs (hunk 4 above)
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Makes sense! Will clarify

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs (hunk 6 above)
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs:297]: Will do

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs (hunk 7 above)
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Will make sure that is covered in unit tests

File: rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs (hunk 5 above)
Comment (drewrelmas) [reply rust/otap-dataflow/crates/core-nodes/src/processors/transform_processor/mod.rs]: Yes, that was my thinking. I initially thought sanitization should happen before emptiness checks because it could itself reduce to an empty batch, but I think it can't turn a non-empty batch into empty. That part can be moved back to the original spot, but I think the output check should happen first to prevent needing to handle a partially dispatched request.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for the PR. I think would be good to add tests for
 - routed-only NACK.
 - mixed empty and non-empty routed outputs
 - an unconfigured route to cover the ACK/NACK and partial-send edge cases
 
It would also be good for the main regression test to use `skip_sanitize_result: true`, matching the issue reproduction.

**Review by albertlockett — APPROVED** [review]
LGTM, thanks @drewrelmas
